rs/tests/system_tests.bzl

```
@@ -11,151 +11,85 @@ load("//bazel:mainnet-icos-images.bzl", "icos_dev_image_download_url", "icos_ima
-            # relative to the $RUNFILES directory.
-            export RUNFILES="$PWD"
-            mkdir "$TEST_TMPDIR/root_env"
-            "$RUNFILES/{test_executable}" \
-              --working-dir "$TEST_TMPDIR" \
-              --group-base-name {group_base_name} \
-              {logs} \
-              {no_summary_report} \
-              {exclude_logs} \
-              "$@" run
+        exec {run_systest} "$@"
         """.format(
-            test_executable = ctx.executable.src.short_path,
-            group_base_name = ctx.label.name,
-            no_summary_report = "--no-summary-report" if ctx.executable.colocated_test_bin != None else "",
-            info_file = ctx.info_file.short_path,
-            logs = "--no-logs" if no_logs else "",
-            exclude_logs = " ".join(["--exclude-logs {pattern}".format(pattern = pattern) for pattern in ctx.attr.exclude_logs]),
+            run_systest = ctx.executable._run_systest.short_path,
         ),
     )
-
-    env = dict(ctx.attr.env.items())
+    data.append(run_test_script_file)
+    data.append(ctx.executable._run_systest)
 
     # Expand Make variables in env vars, with runtime_deps as targets
     for key, value in env.items():
         # If this looks like a Make variable, try to expand it
         if value.startswith("$"):
             env[key] = ctx.expand_location(value, ctx.attr.runtime_deps)
 
-    env |= {
-        "FARM_METADATA_PATH": ctx.info_file.short_path,
-    }
+    # Used by the run script and by farm to read metadata
+    env["FARM_METADATA_PATH"] = ctx.info_file.short_path
+    data.append(ctx.info_file)
 
     # We use the RUN_SCRIPT_ prefix for variables that are processed by the run
     # script, and not passed directly to the test.
 
+    # The test driver
+    env["RUN_SCRIPT_TEST_EXECUTABLE"] = ctx.executable.src.short_path
+    data.append(ctx.executable.src)
+
+    # The uploader for dependencies fetched remotely from the runner
+    env["RUN_SCRIPT_UPLOAD_SYSTEST_DEP"] = ctx.executable._upload_systest_dep.short_path
+    data.append(ctx.executable._upload_systest_dep)
+
     # RUN_SCRIPT_ICOS_IMAGES:
     # Have the run script resolve repo based ICOS images.
     # The run script expects a map of enviromment variable prefixes to targets. e.g.
     # RUN_SCRIPT_ICOS_IMAGES=ENV_DEPS__GUESTOS_DISK_IMG:ic-os/guestos/envs/dev/disk-img.tar.zst;ENV_DEPS__GUESTOS_UPDATE_IMG:ic-os/guestos/envs/dev/update-img.tar.zst
-    icos_images = ctx.attr.icos_images
-    env |= {
-        "RUN_SCRIPT_ICOS_IMAGES": ";".join([k + ":" + v.files.to_list()[0].short_path for k, v in icos_images.items()]),
-    }
-
-    env["RUN_SCRIPT_UPLOAD_SYSTEST_DEP"] = ctx.executable._upload_systest_dep.short_path
+    env["RUN_SCRIPT_ICOS_IMAGES"] = ";".join([k + ":" + v.files.to_list()[0].short_path for k, v in ctx.attr.icos_images.items()])
+    [data.append(image.files.to_list()[0]) for _, image in ctx.attr.icos_images.items()]
```

You could also consider the slightly shorter:
```suggestion
    data += [image.files.to_list()[0] for _, image in ctx.attr.icos_images.items()]
```

---

I never know what starlark supports and what it doesn't 😛 I prefer your version, thanks!

---

Ah, I had to push some conflicting changes to appease gazelle. Will use your solution instead